### PR TITLE
feat: Added ability to specify KMS, SSE and SSEC options when either gettin…

### PR DESCRIPTION
**Pull request by aaronmell**
…g or putting an item in a bucket, and also when creating a bucket

Signed-off-by: Aaron Mell aaronamell@gmail.com

### CODE REVIEW

**Comment by aaronmell** [comment]
Our organization requires that all s3 buckets use encryption. I need this for workflows. I'd like feedback on this, while I am working on adding the required functionality to workflow

Relevant workflow issue
https://github.com/argoproj/argo-workflows/issues/2800

File: s3/s3.go [hunk 1]
```diff
@@ -55,6 +66,7 @@ type S3ClientOpts struct {
 	RoleARN         string
 	RoleSessionName string
 	UseSDKCreds     bool
+	EncryptOpts     *EncryptOpts
```
Comment (alexec) [inline s3/s3.go:69]: minor - if this was non-pointer - you could avoid nil checks

File: s3/s3.go [hunk 2]
```diff
@@ -296,3 +340,80 @@ func IsS3ErrCode(err error, code string) bool {
...
 	return false
 }
+
+// setBucketEnc sets the encryption options on a bucket
+func (s *s3client) setBucketEnc(bucketName string) error {
+	if s.EncryptOpts == nil || !s.EncryptOpts.Enabled {
+		return nil
+	}
+
+	log.Infof("Setting Bucket Encryption: %s. (KmsKeyId: %s)", bucketName, s.EncryptOpts.KmsKeyId)
```
Comment (alexec) [inline s3/s3.go:350]: should use structured logging

File: s3/s3.go [hunk 3]
```diff
@@ -296,3 +340,80 @@ func IsS3ErrCode(err error, code string) bool {
...
+	var config *sse.Configuration
+
+	if s.EncryptOpts.KmsKeyId != "" {
+		config = sse.NewConfigurationSSEKMS(s.EncryptOpts.KmsKeyId)
+	} else {
+		config = sse.NewConfigurationSSES3()
+	}
+
+	err := s.minioClient.SetBucketEncryption(s.ctx, bucketName, config)
+	return errors.WithStack(err)
```
Comment (alexec) [inline s3/s3.go:360]: no need to use `WithStack`

File: s3/s3.go [hunk 4]
```diff
@@ -296,3 +340,80 @@ func IsS3ErrCode(err error, code string) bool {
...
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to parse KMS encryption context")
+		}
+
+		if encryptionCtx == nil {
+			// To overcome a limitation in Minio which checks interface{} == nil.
+			kms, err := encrypt.NewSSEKMS(e.KmsKeyId, nil)
+
+			if err != nil {
+				return nil, errors.WithStack(err)
```
Comment (alexec) [inline s3/s3.go:387]: as above, `WithStack` makes code more verbose...

File: s3/s3.go (hunk 2 above)
Comment (aaronmell) [reply s3/s3.go:350]: Do you have an example of structure logging? I just copied that from other places in the code that follow a similar pattern.

File: s3/s3.go (hunk 1 above)
Comment (aaronmell) [reply s3/s3.go:69]: Should it be optional? I don't have a good grasp on how this project is used by the other argo projects. If its non-optional, that would require downstream projects to fix their code when they update the library right? Would I need to update all the downstream projects?

File: s3/s3.go (hunk 2 above)
Comment (aaronmell) [reply s3/s3.go]: Found a pattern, went ahead and refactored the other log statements as well.

**Comment by aaronmell** [comment]
I've tested everything in workflow. I'll be submitting a PR for workflow later this afternoon or tomorrow morning

File: s3/s3.go (hunk 1 above)
Comment (alexec) [reply s3/s3.go:69]: If it were non pointer, then enabled would be the zero value, false.

Comment (aaronmell) [reply s3/s3.go]: Updated

**Review by alexec — APPROVED** [review]

**Comment by alexec** [comment]
v0.11.0 tagged.

**Pull request merged**